### ISSUE TITLE
Puma unresponsive when there is more than one ActionCable channel

### ISSUE THREAD

**Issue by robertlyall** [rails/rails#25117]
### Steps to reproduce

I have made a repository detailing the issue here: 

https://github.com/robertlyall/puma_channel_problem

If you checkout `879fde` (one channel), and start the server - everything works wonderfully. Shutdown the server then checkout `edc493` (two channels). After a number of page refreshes the request will hang. If you run `netstat -atn | grep 3000` you'll see something like this:

```
tcp4    1320      0  127.0.0.1.3000         127.0.0.1.57557        ESTABLISHED
tcp4    1320      0  127.0.0.1.3000         127.0.0.1.57553        CLOSE_WAIT
tcp4    1320      0  127.0.0.1.3000         127.0.0.1.57549        CLOSE_WAIT
tcp4    1320      0  127.0.0.1.3000         127.0.0.1.57547        CLOSE_WAIT
tcp4    1320      0  127.0.0.1.3000         127.0.0.1.57541        CLOSE_WAIT
tcp4    1320      0  127.0.0.1.3000         127.0.0.1.57538        CLOSE_WAIT
tcp4    1152      0  127.0.0.1.3000         127.0.0.1.57536        CLOSE_WAIT
tcp4    1320      0  127.0.0.1.3000         127.0.0.1.57534        CLOSE_WAIT
tcp4    1320      0  127.0.0.1.3000         127.0.0.1.57532        CLOSE_WAIT
tcp4    1320      0  127.0.0.1.3000         127.0.0.1.57530        CLOSE_WAIT
tcp4    1320      0  127.0.0.1.3000         127.0.0.1.57528        CLOSE_WAIT
tcp4       0      0  127.0.0.1.3000         127.0.0.1.57525        CLOSE_WAIT
tcp4       0      0  127.0.0.1.3000         127.0.0.1.57523        CLOSE_WAIT
tcp4       0      0  127.0.0.1.3000         127.0.0.1.57520        ESTABLISHEDD
tcp4       0      0  127.0.0.1.3000         127.0.0.1.57516        CLOSE_WAIT
```

### Expected behavior

Shouldn't hang if more than one channel is being used.

### Actual behavior

Puma server hangs in development. Status for each thread/worker is set to `CLOSE_WAIT`. Have to kill the processes manually. 

### System configuration

**Rails version**:

5.0.0.rc1

**Ruby version**:

ruby 2.3.0p0 (2015-12-25 revision 53290) [x86_64-darwin15]

I've tried this on three different machines (two Mac, one linux) and the problem occurs on all three. Please let me know if there's any more information you require from me.

**Comment by jeremy** [comment]
Thanks for the repro app, @robertlyall!

Are you seeing this in development mode only, or production as well?

**Comment by robertlyall** [comment]
Only in development as far as I can tell. I spent the last few minutes trying to replicate in production without any luck. Was the issue reproducible for you using my repo app? Just want to clarify it's not just me.

**Comment by nthj** [comment]
I've also been able to reproduce all day in development only.  OS X Yosemite.

``` ruby -v
ruby 2.3.0p0 (2015-12-25 revision 53290) [x86_64-darwin14]

rails -v
Rails 5.0.0.rc1```

It only seemed to happen when I had multiple channels.

**Comment by anykeyh** [comment]
I guess it's related, cf. the issue I've opened in the puma's repos:

Hello, 

I've a strange issue happening for now in development (not sure about production yet).

The step to reproduce are simple:

- Launch the server (in my case using `rails s`)
- Open a website using actioncable with multiple channels
- Stop the server and launch it again
- Wait for auto-reconnect of action cable

And TADA, broken :(

Here what is inside my console:

```
[44646] Puma starting in cluster mode...
[44646] * Version 3.4.0 (ruby 2.3.0-p0), codename: Owl Bowl Brawl
[44646] * Min threads: 1, max threads: 8
[44646] * Environment: development
[44646] * Process workers: 1
[44646] * Phased restart available
[44646] * Listening on tcp://localhost:3000
[44646] Use Ctrl-C to stop
[44646] * Starting control server on unix:///var/folders/sh/64tpb9ys67l6smwkqlmhz5pr0000gn/T/puma-status-1462691832716-44646
[44646] - Worker 0 (pid: 44649) booted, phase: 0
  ActiveRecord::SchemaMigration Load (0.5ms)  SELECT "schema_migrations".* FROM "schema_migrations"
Started GET "/live/" [WebSocket] for ::1 at 2016-05-08 15:17:14 +0800
Successfully upgraded to WebSocket (REQUEST_METHOD: GET, HTTP_CONNECTION: Upgrade, HTTP_UPGRADE: websocket)
Registered connection (08657a35-aca7-404c-b44c-84ca3793d07c)
^C[44646] - Gracefully shutting down workers...
```

Note the ^C (Try to cancel, can spam nothing will happens)

The only way for me to kill is to use `kill -9 PID`, and I can't get any stacktrace.

**Comment by ethayer** [comment]
FWIW- here is my expirence:
Getting rid of gem 'web-console'  then running...
```bundle exec puma```
seems to work for me.

**questionable magic:**
If you then run ```bin/rails s``` directly after shutting down ```bundle exec puma```
rails server runs fine with running logs...
then once you shut down the server and run ```bin/rails s``` directly after
server then hangs.

**Comment by AndreasWurm** [comment]
see #24741

**Comment by robertlyall** [comment]
@AndreasWurm: Multiple channels work with the proposed patch, however I'm getting errors in my development log.

```
NoMethodError (undefined method `[]' for nil:NilClass):

/Users/robertlyall/Downloads/rails-5.0.0.rc1/actionpack/lib/action_dispatch/journey/router.rb:41:in `block in serve'
/Users/robertlyall/Downloads/rails-5.0.0.rc1/actionpack/lib/action_dispatch/journey/router.rb:26:in `each'
/Users/robertlyall/Downloads/rails-5.0.0.rc1/actionpack/lib/action_dispatch/journey/router.rb:26:in `serve'
/Users/robertlyall/Downloads/rails-5.0.0.rc1/actionpack/lib/action_dispatch/routing/route_set.rb:725:in `call'
/Users/robertlyall/Downloads/rails-5.0.0.rc1/actionview/lib/action_view/digestor.rb:12:in `call'
rack (2.0.0.rc1) lib/rack/etag.rb:25:in `call'
rack (2.0.0.rc1) lib/rack/conditional_get.rb:25:in `call'
rack (2.0.0.rc1) lib/rack/head.rb:12:in `call'
rack (2.0.0.rc1) lib/rack/session/abstract/id.rb:222:in `context'
rack (2.0.0.rc1) lib/rack/session/abstract/id.rb:216:in `call'
/Users/robertlyall/Downloads/rails-5.0.0.rc1/actionpack/lib/action_dispatch/middleware/cookies.rb:613:in `call'
/Users/robertlyall/Downloads/rails-5.0.0.rc1/activerecord/lib/active_record/migration.rb:552:in `call'
/Users/robertlyall/Downloads/rails-5.0.0.rc1/actionpack/lib/action_dispatch/middleware/callbacks.rb:38:in `block in call'
/Users/robertlyall/Downloads/rails-5.0.0.rc1/activesupport/lib/active_support/callbacks.rb:97:in `__run_callbacks__'
/Users/robertlyall/Downloads/rails-5.0.0.rc1/activesupport/lib/active_support/callbacks.rb:750:in `_run_call_callbacks'
/Users/robertlyall/Downloads/rails-5.0.0.rc1/activesupport/lib/active_support/callbacks.rb:90:in `run_callbacks'
/Users/robertlyall/Downloads/rails-5.0.0.rc1/actionpack/lib/action_dispatch/middleware/callbacks.rb:36:in `call'
/Users/robertlyall/Downloads/rails-5.0.0.rc1/actionpack/lib/action_dispatch/middleware/executor.rb:12:in `call'
/Users/robertlyall/Downloads/rails-5.0.0.rc1/actionpack/lib/action_dispatch/middleware/remote_ip.rb:79:in `call'
/Users/robertlyall/Downloads/rails-5.0.0.rc1/actionpack/lib/action_dispatch/middleware/debug_exceptions.rb:49:in `call'
web-console (3.2.0) lib/web_console/middleware.rb:131:in `call_app'
web-console (3.2.0) lib/web_console/middleware.rb:28:in `block in call'
web-console (3.2.0) lib/web_console/middleware.rb:18:in `catch'
web-console (3.2.0) lib/web_console/middleware.rb:18:in `call'
/Users/robertlyall/Downloads/rails-5.0.0.rc1/actionpack/lib/action_dispatch/middleware/show_exceptions.rb:31:in `call'
/Users/robertlyall/Downloads/rails-5.0.0.rc1/railties/lib/rails/rack/logger.rb:36:in `call_app'
/Users/robertlyall/Downloads/rails-5.0.0.rc1/railties/lib/rails/rack/logger.rb:24:in `block in call'
/Users/robertlyall/Downloads/rails-5.0.0.rc1/activesupport/lib/active_support/tagged_logging.rb:70:in `block in tagged'
/Users/robertlyall/Downloads/rails-5.0.0.rc1/activesupport/lib/active_support/tagged_logging.rb:26:in `tagged'
/Users/robertlyall/Downloads/rails-5.0.0.rc1/activesupport/lib/active_support/tagged_logging.rb:70:in `tagged'
/Users/robertlyall/Downloads/rails-5.0.0.rc1/railties/lib/rails/rack/logger.rb:24:in `call'
/Users/robertlyall/Downloads/rails-5.0.0.rc1/actionpack/lib/action_dispatch/middleware/request_id.rb:24:in `call'
rack (2.0.0.rc1) lib/rack/method_override.rb:22:in `call'
rack (2.0.0.rc1) lib/rack/runtime.rb:22:in `call'
/Users/robertlyall/Downloads/rails-5.0.0.rc1/activesupport/lib/active_support/cache/strategy/local_cache_middleware.rb:28:in `call'
/Users/robertlyall/Downloads/rails-5.0.0.rc1/actionpack/lib/action_dispatch/middleware/executor.rb:12:in `call'
/Users/robertlyall/Downloads/rails-5.0.0.rc1/actionpack/lib/action_dispatch/middleware/static.rb:136:in `call'
rack (2.0.0.rc1) lib/rack/sendfile.rb:111:in `call'
/Users/robertlyall/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/bundler/gems/raven-ruby-a49f486530db/lib/raven/integrations/rack.rb:53:in `call'
/Users/robertlyall/Downloads/rails-5.0.0.rc1/railties/lib/rails/engine.rb:522:in `call'
puma (3.4.0) lib/puma/configuration.rb:224:in `call'
puma (3.4.0) lib/puma/server.rb:569:in `handle_request'
puma (3.4.0) lib/puma/server.rb:406:in `process_client'
puma (3.4.0) lib/puma/server.rb:271:in `block in run'
puma (3.4.0) lib/puma/thread_pool.rb:114:in `block in spawn_thread'
  Authie::Session Load (21.4ms)  SELECT  "authie_sessions".* FROM "authie_sessions" WHERE "authie_sessions"."active" = $1 AND "authie_sessions"."token" = $2 ORDER BY "authie_sessions"."id" ASC LIMIT $3  [["active", true], ["token", "vYyEk0yy9zG/WUXv68YsbW7YaBojajx2gv8o7yQgJ0Q="], ["LIMIT", 1]]
  User Load (0.7ms)  SELECT  "users".* FROM "users" WHERE "users"."id_for_authie" = $1 LIMIT $2  [["id_for_authie", 250147961], ["LIMIT", 1]]
[ActionCable] [Robert Lyall] Registered connection (Z2lkOi8vam9iYmJzL1VzZXIvN2I0NjkzMDQtZmUzOC00YmYxLWFhODItMTI0ZTJlMjZlNTc1)
  Rendering /Users/robertlyall/Downloads/rails-5.0.0.rc1/actionpack/lib/action_dispatch/middleware/templates/rescues/diagnostics.html.erb within rescues/layout
  Rendering /Users/robertlyall/Downloads/rails-5.0.0.rc1/actionpack/lib/action_dispatch/middleware/templates/rescues/_source.html.erb
  Rendered /Users/robertlyall/Downloads/rails-5.0.0.rc1/actionpack/lib/action_dispatch/middleware/templates/rescues/_source.html.erb (2.6ms)
  Rendering /Users/robertlyall/Downloads/rails-5.0.0.rc1/actionpack/lib/action_dispatch/middleware/templates/rescues/_trace.html.erb
  Rendered /Users/robertlyall/Downloads/rails-5.0.0.rc1/actionpack/lib/action_dispatch/middleware/templates/rescues/_trace.html.erb (1.9ms)
  Rendering /Users/robertlyall/Downloads/rails-5.0.0.rc1/actionpack/lib/action_dispatch/middleware/templates/rescues/_request_and_response.html.erb
  Rendered /Users/robertlyall/Downloads/rails-5.0.0.rc1/actionpack/lib/action_dispatch/middleware/templates/rescues/_request_and_response.html.erb (1.3ms)
  Rendered /Users/robertlyall/Downloads/rails-5.0.0.rc1/actionpack/lib/action_dispatch/middleware/templates/rescues/diagnostics.html.erb within rescues/layout (64.4ms)
[ActionCable] [Robert Lyall] UserChannel is transmitting the subscription confirmation
[ActionCable] [Robert Lyall] UserChannel is streaming from user_room_7b469304-fe38-4bf1-aa82-124e2e26e575
[ActionCable] [Robert Lyall] AccountChannel is transmitting the subscription confirmation
[ActionCable] [Robert Lyall] AccountChannel is streaming from account_room_ee7bc464-2534-4fb2-b3ce-6b9b9efba07e
```

**Comment by machty** [comment]
I'm still running into deadlocking with multiple channels and I'm on 5.0.0.rc1.

I've found that if I set `config.eager_load = true` in development, the problem goes away (and since this is the default on prod, this explains why other people haven't been seeing this issue on prod).

Also, if I put code on the client side to put a delay between the channel subscriptions, the problem also seems to go away.
